lib/Sema/TypeCheckType.cpp

```diff
@@ -1051,22 +1055,9 @@ static Type SelfAllowedBySE0068(TypeResolution resolution,
       (nominal = nominalDC->getSelfNominalTypeDecl())) {
     assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
 
-    bool insideClass = nominalDC->getSelfClassDecl() != nullptr;
-    AbstractFunctionDecl *methodDecl = dc->getInnermostMethodContext();
-    bool declaringMethod = methodDecl &&
-      methodDecl->getDeclContext() == dc->getParentForLookup();
-    bool isMutablePropertyOrSubscriptOfClass = insideClass &&
-      (options.is(TypeResolverContext::PatternBindingDecl) ||
-       options.is(TypeResolverContext::FunctionResult));
-    bool isTypeAliasInClass = insideClass &&
-      options.is(TypeResolverContext::TypeAliasDecl);
-
-    if (((!insideClass || !declaringMethod) &&
-         !isMutablePropertyOrSubscriptOfClass && !isTypeAliasInClass &&
-         !options.is(TypeResolverContext::GenericRequirement)) ||
-        options.is(TypeResolverContext::ExplicitCastExpr)) {
+    if (!options.is(TypeResolverContext::GenericRequirement)) {
       Type SelfType = nominal->getSelfInterfaceType();
-      if (insideClass)
+      if (nominalDC->getSelfClassDecl() != nullptr)
```

Comment:
Just a final idle thought. Should this be:
```
      if (nominalDC->getSelfClassDecl() && !nominalDC->getSelfClassDecl()->isFinal())
```

Reply:
It should resolve this problem: https://github.com/apple/swift/pull/22863#issuecomment-473601928